flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/InputBuffersMetricsTest.java

```diff
@@ -0,0 +1,293 @@
+			numberOfLocalChannelsGate2).f0;
+
+		SingleInputGate inputGate1 = tuple1.f0;
+
+		RemoteInputChannel remoteInputChannel1 = tuple1.f1.get(0);
+
+		SingleInputGate[] inputGates = new SingleInputGate[]{tuple1.f0, inputGate2};
+		FloatingBuffersUsageGauge floatingBuffersUsageGauge = new FloatingBuffersUsageGauge(inputGates);
+		ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge = new ExclusiveBuffersUsageGauge(inputGates);
+		CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge = new CreditBasedInputBuffersUsageGauge(
+			floatingBuffersUsageGauge,
+			exclusiveBuffersUsageGauge,
+			inputGates);
+
+		try (CloseableRegistry closeableRegistry = new CloseableRegistry()) {
+			assertEquals(0.0, floatingBuffersUsageGauge.getValue(), 0.0);
+			assertEquals(0.0, inputBuffersUsageGauge.getValue(), 0.0);
+
+			// drain gate1's exclusive buffers
+			drainBuffer(buffersPerChannel, remoteInputChannel1, closeableRegistry);
+
+			int totalBuffers = extraNetworkBuffersPerGate * inputGates.length + buffersPerChannel * totalNumberOfRemoteChannels;
+
+			remoteInputChannel1.requestSubpartition(0);
+
+			int backlog = 3;
+			int totalRequestedBuffers = buffersPerChannel + backlog;
+
+			remoteInputChannel1.onSenderBacklog(backlog);
+
+			assertEquals(totalRequestedBuffers, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+
+			drainBuffer(totalRequestedBuffers, remoteInputChannel1, closeableRegistry);
+
+			assertEquals(0, remoteInputChannel1.unsynchronizedGetFloatingBuffersAvailable());
+			assertEquals((double) (buffersPerChannel + totalRequestedBuffers) / totalBuffers,
+				inputBuffersUsageGauge.getValue(), 0.0001);
+		} finally {
+			inputGate1.close();
+			inputGate2.close();
+			network.close();
+		}
+	}
+
+	private void drainAndValidate(
+		int numBuffersToRequest,
+		int totalRequestedBuffers,
+		RemoteInputChannel channel,
+		CloseableRegistry closeableRegistry,
+		int totalBuffers,
+		int totalExclusiveBuffers,
+		ExclusiveBuffersUsageGauge exclusiveBuffersUsageGauge,
+		CreditBasedInputBuffersUsageGauge inputBuffersUsageGauge,
+		SingleInputGate inputGate) throws IOException {
+
+		drainBuffer(numBuffersToRequest, channel, closeableRegistry);
+		assertEquals(totalRequestedBuffers, exclusiveBuffersUsageGauge.calculateUsedBuffers(inputGate));
+		assertEquals((double) totalRequestedBuffers / totalExclusiveBuffers, exclusiveBuffersUsageGauge.getValue(), 0.0001);
+		assertEquals((double) totalRequestedBuffers / totalBuffers, inputBuffersUsageGauge.getValue(), 0.0001);
+	}
+
+	private void drainBuffer(int boundary, RemoteInputChannel channel, CloseableRegistry closeableRegistry) throws IOException {
+		for (int i = 0; i < boundary; i++) {
+			Buffer buffer = channel.requestBuffer();
+			if (buffer != null) {
+				closeableRegistry.registerCloseable(buffer::recycleBuffer);
+			} else {
+				break;
+			}
+		}
+	}
+
+	private Tuple3<SingleInputGate, List<RemoteInputChannel>, List<LocalInputChannel>> buildInputGate(
```

Comment:
we could only return `Tuple2<SingleInputGate, List<RemoteInputChannel>>` instead because the list of local channels would not be used any more.

Reply:
yes